src/Symfony/Component/HttpKernel/HttpCache/ResponseCacheStrategy.php

```diff
@@ -64,33 +99,116 @@ public function update(Response $response)
             return;
         }
 
-        // Remove validation related headers in order to avoid browsers using
-        // their own cache, because some of the response content comes from
-        // at least one embedded response (which likely has a different caching strategy).
-        if ($response->isValidateable()) {
-            $response->setEtag(null);
-            $response->setLastModified(null);
+        // Remove validation related headers of the master response,
+        // because some of the response content comes from at least
+        // one embedded response (which likely has a different caching strategy).
+        $response->setEtag(null);
+        $response->setLastModified(null);
+
+        $this->add($response);
+
+        $response->headers->set('Age', $this->age);
+
+        if ($this->isNotCacheableResponseEmbedded) {
+            $response->setExpires($response->getDate());
+
+            if ($this->flagDirectives['no-store']) {
+                $response->headers->set('Cache-Control', 'no-cache, no-store, must-revalidate');
+            } else {
+                $response->headers->set('Cache-Control', 'no-cache, must-revalidate');
+            }
+
+            return;
+        }
+
+        $flags = array_filter($this->flagDirectives);
+
+        if (isset($flags['must-revalidate'])) {
+            $flags['no-cache'] = true;
         }
 
-        if (!$response->isFresh() || !$response->isCacheable()) {
-            $this->cacheable = false;
+        $response->headers->set('Cache-Control', implode(', ', array_keys($flags)));
+
+        $maxAge = null;
+        $sMaxage = null;
+
+        if ($this->ageDirectives['max-age']) {
+            $maxAge = $this->ageDirectives['max-age'] + $this->age;
+            $response->headers->addCacheControlDirective('max-age', $maxAge);
         }
 
-        if (!$this->cacheable) {
-            $response->headers->set('Cache-Control', 'no-cache, must-revalidate');
+        if ($this->ageDirectives['s-maxage']) {
+            $sMaxage = $this->ageDirectives['s-maxage'] + $this->age;
 
-            return;
+            if ($maxAge !== $sMaxage) {
```

Comment:
that seems unexpected to me. while it is somewhat redundant, i don't see symfony doing the same in a normal response, so i'd not eliminate the duplication.

Reply:
The duplication is produced by the `ResponseCacheStrategy` itself, because we need to merge `max-age` into `s-maxage` if only the former is set, because another response might only have the later and then `s-maxage` could still be sent.

Reply:
we could try to track things separately. but i think in the end it would be good enough to check if the master response has s-maxage set and in that case re-add it regardless?

oh, i just notice that you do not remove s-maxage from the master response, only overwrite it if its different from max-age. could that be a problem when both are 0?

Reply:
All existing Cache-Control headers on the master response are overwritten by https://github.com/symfony/symfony/pull/26532/files/77ae8bc10240721c44a78e3338d5c28dd7c11c55#diff-e5a339b48cec0fa0a4d0c5c4c705f568R130. `s-maxage` is only added if it differs from `max-age`. Be aware that the master response is no difference than any ESI response, it is first passed to `update` so it's Cache-Control values are in the ResponseCacheStrategy already.

Reply:
oh, indeed, missed that. then could we track if the master response has an s-maxage set, and if so keep setting it regardless of whether it is different from max-age, to avoid a surprise?

Reply:
We can't set a `s-maxage` of the master response, if the ESI responses do not have that. We can only set a Cache-Control header if **all** responses have a value, and if we can merge them.

Reply:
right. there is still the edge case when master response and all fragments have s-maxage == max-age. then we would not set the (redundant) s-maxage because its the same as max-age. this might be a bit confusing, but if "phantom" s-maxage appears its also confusing, so i guess we can ignore the issue.

Reply:
genau. i agreed on leaving it as is, even if there is an edge case that could be a bit confusing, where you set both values on all fragments and the master response, but end up having only max-age.